src/transport_ucx.h

```diff
@@ -0,0 +1,846 @@
+            status = ucp_request_check_status(req);
+        } while (status == UCS_INPROGRESS);
+        ucp_request_release(req);
+        return status;
+    }
+}
+
+static inline
+ucs_status_t shmem_transport_ucx_release_op(ucs_status_ptr_t req) {
+    if (req == NULL)
+         return UCS_OK;
+    else if (UCS_PTR_IS_ERR(req))
+        return UCS_PTR_STATUS(req);
+    else {
+        ucp_request_free(req);
+        return UCS_INPROGRESS;
+    }
+}
+
+static inline
+ucs_status_t shmem_transport_ucx_post_cb_op(ucs_status_ptr_t req, void *completion) {
+    if (req == NULL) {
+        __atomic_store_n((long*)completion, 1, __ATOMIC_RELEASE);
+        return UCS_OK;
+    } else if (UCS_PTR_IS_ERR(req))
+        return UCS_PTR_STATUS(req);
+    else {
+        shmem_transport_ucx_req_t *preq = (shmem_transport_ucx_req_t *) req;
+
+        /* Populate the request and set the valid flag. The valid flag resolves
+         * a race with the completion callback, which could occur in a separate
+         * thread. */
+        preq->ptr = completion;
+        __atomic_store_n(&preq->valid, 1, __ATOMIC_RELEASE);
+
+        /* Request is released instead of freed because we expect a completion callback */
+        ucp_request_release(req);
+        return UCS_INPROGRESS;
+    }
+}
+
+static inline
+void shmem_transport_ucx_get_mr(const void *addr, int dest_pe,
+                                uint8_t **remote_addr, ucp_rkey_h *rkey) {
+    if ((void*) addr >= shmem_internal_data_base &&
+        (uint8_t*) addr < (uint8_t*) shmem_internal_data_base + shmem_internal_data_length) {
+
+        *rkey = shmem_transport_peers[dest_pe].data_rkey;
+#ifdef ENABLE_REMOTE_VIRTUAL_ADDRESSING
+        *remote_addr = (uint8_t *) addr;
+#else
+        *remote_addr = (uint8_t*) (((uint8_t *) addr - (uint8_t *) shmem_internal_data_base) +
+                       shmem_transport_peers[dest_pe].data_base);
+#endif
+    } else if ((void*) addr >= shmem_internal_heap_base &&
+               (uint8_t*) addr < (uint8_t*) shmem_internal_heap_base + shmem_internal_heap_length) {
+
+        *rkey = shmem_transport_peers[dest_pe].heap_rkey;
+#ifdef ENABLE_REMOTE_VIRTUAL_ADDRESSING
+        *remote_addr = (uint8_t *) addr;
+#else
+        *remote_addr = (uint8_t*) (((uint8_t *) addr - (uint8_t *) shmem_internal_heap_base) +
+                       shmem_transport_peers[dest_pe].heap_base);
+#endif
+    } else {
+        RAISE_ERROR_MSG("address (%p) outside of symmetric areas\n", addr);
+    }
+}
+
+static inline
+int
+shmem_transport_ctx_create(struct shmem_internal_team_t *team, long options, shmem_transport_ctx_t **ctx)
+{
+    /* FIXME: Contexts are not optimized yet. We could create a separate set of
+     * EPs for each context (or team), and/or we could use completion callbacks
+     * to implement completion counters at the SOS level to enable separate
+     * completion tracking per context. */
+
+    if (team == NULL)
+        return 1;
```

Comment:
Sorry if this is repetitive, but this is slightly different than OFI and Portals. In those cases, we raise error and abort. In this case, we are returning without error. In #941, I am proposing a check for the return value from the transport layer and assigning `SHMEMX_CTX_INVALID` in case of non-zero return value. I think we may need to unify these two behaviors, any of them should be fine. @davidozog @jdinan Your thoughts?

Reply:
I'll update to match OFI and Portals. IIUC, team handle checks should have been resolved at the upper layer and it's an error if NULL reaches the transport layer.